[WIP] upgrade reanimated to v3

OmarBasem opened the pull request:
fixes: https://github.com/status-im/status-mobile/issues/15764

OmarBasem committed:
upgrade reanimated to v3

siddarthkay commented on nix/deps/gradle/proj.list:
hmm something went wrong over here, we should revert this file to the way it was

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :x: | 697e9390 | [#1](https://ci.status.im/job/status-mobile/job/prs/job/android/job/PR-15769/1/) | 2023-04-27 13:06:04 | ~2 min | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-mobile/job/prs/job/android/job/PR-15769/1/consoleText) |

OmarBasem committed:
revert file

OmarBasem commented:
When running `make run-android` I get the following error:

```
> java.lang.RuntimeException: com.android.builder.sdk.InstallFailedException: Failed to install the following SDK components:
      patcher;v4 SDK Patch Applier v4
      ndk;23.1.7779620 NDK (Side by side) 23.1.7779620
```

Upgrading NDK to version `23.1.7779620` in `nix/pkgs/android-sdk/compose.nix` gives the following error:

``` /nix/store/m4h36k1cbmhavv177mgmyg4q8vjri7b3-gomobile-unstable-2022-05-18/bin/gomobile: no usable NDK in /nix/store/n4zz2v9vadj706rbpbpa6f870ix5vl7j-androidsdk-mod-sdk: open /nix/store/n4zz2v9vadj706rbpbpa6f870ix5vl7j-androidsdk-mod-sdk/ndk/23.1.7779620/meta/platforms.json: no such file or directory, open /nix/store/n4zz2v9vadj706rbpbpa6f870ix5vl7j-androidsdk-mod-sdk/ndk-bundle/meta/platforms.json: no such file or directory```


Also, running `make android-clean` gives the following error:

``` Configure project :react-native-reanimated
No AAR for react-native-reanimated found. Attempting to build from source.
Android gradle plugin: 7.3.1
Gradle: 7.5
WARNING:Using flatDir should be avoided because it doesn't support any meta-data formats.
WARNING:Software Components will not be created automatically for Maven publishing from Android Gradle Plugin 8.0. To opt-in to the future behavior, set the Gradle property android.disableAutomaticComponentCreation=true in the `gradle.properties` file or use the new publishing DSL.
```

jakubgs commented:
It doesn't seem like you changed the SDK version:
https://github.com/status-im/status-mobile/blob/fb5caa7064e824cc7de950f668d0ca321f1f6a0c/nix/pkgs/android-sdk/compose.nix#L16
If it even is available in our version of nixpkgs.

OmarBasem commented:
@jakubgs I tried changing it, and then reverted it back

OmarBasem committed:
update ndk version

OmarBasem commented:
I just push a commit with the updated version to see if it will be any different on CI.

I see the NDK version in `/nix/pkgs/android-sdk/compose.nix` do not match the ones in `/android/build.gradle`

jakubgs commented:
The AAPT2 build issue is known, and was fixed in one of the RN update PRs with this: https://github.com/status-im/status-mobile/commit/d69277ed

jakubgs commented:
Yes, value in `android/build.gradle` should normally reflect the version we are using from Nix.

Also, please don't add a root slash to relative paths.
It's `android/build.gradle`, not `/android/build.gradle`, pretty sure it's not at the root of your filesystem.

OmarBasem committed:
update default.nix

OmarBasem commented:
> The AAPT2 build issue is known, and was fixed in one of the RN update PRs with this: [d69277ed](https://github.com/status-im/status-mobile/commit/d69277ed)

@jakubgs after doing that update to `default.nix` I am getting the following error:
`error: 1 dependencies of derivation '/nix/store/vhknay6drrb228a7hz0hpy3s4kjb0jv0-status-mobile-maven-deps.drv' failed to build
error: build of '/nix/store/1wvicqsq8p6kiqidibd36gd4pn25ajyf-status-go-0.148.3-8608aec-android.drv', '/nix/store/vhknay6drrb228a7hz0hpy3s4kjb0jv0-status-mobile-maven-deps.drv' failed
make: *** [run-android] Error 1`

jakubgs commented:
No, that's not the error. That's just what is printed after the actual error.
Your builds are failing due to missing gradle deps:
```
> Could not resolve all files for configuration ':classpath'.
   > Could not find commons-codec:commons-codec:1.11.
     Searched in the following locations:
       - file:/build/android/libs/commons-codec-1.11.jar
       - file:/build/android/libs/commons-codec.jar
       - file:/build/android/app/libs/commons-codec-1.11.jar
       - file:/build/android/app/libs/commons-codec.jar
       - file:/nix/store/csq035dx5q64pxjnxdn8x4h0ap2b14s3-status-mobile-maven-deps/commons-codec/commons-codec/1.11/commons-codec-1.11.pom
     Required by:
         project : > com.android.tools.build:gradle:7.3.1 > com.android.tools.build:builder:7.3.1
```
Are you sure you updated them?

jakubgs committed:
add org.jetbrains.kotlin:kotlin-gradle-plugin:1.5.20

Signed-off-by: Jakub Sokołowski <jakub@status.im>

jakubgs commented:
Oh, I see, running `make nix-update-gradle` is failing with:
```
* What went wrong:
The Android Gradle plugin supports only Kotlin Gradle plugin version 1.5.20 and higher.
The following dependencies do not satisfy the required version:
root project 'StatusIm' -> org.jetbrains.kotlin:kotlin-gradle-plugin:1.4.31
```
That's what the `nix/deps/gradle/add_package.sh` script was created for:
```
 > make shell TARGET=gradle
Configuring Nix shell for target 'gradle'...
Yarn modules changed, copying new version over
Copying node_modules from Nix store:
 - /nix/store/zbrx7vqcz1y3r60md3qricgv2bp7x72b-status-mobile-node-deps-1.22.0
Done in: 12.136s

[nix-shell:~/work/status-mobile]$ nix/deps/gradle/add_package.sh org.jetbrains.kotlin:kotlin-gradle-plugin:1.5.20
Changes made:
 nix/deps/gradle/deps.urls | 27 +++++++++++++++++++++++++++
 1 file changed, 27 insertions(+)

Regenerating Nix files...
Successfully added: org.jetbrains.kotlin:kotlin-gradle-plugin:1.5.20

NOTICE: Running 'make nix-update-gradle' in a new shell is recommended.
```
I've pushed a commit with updated dependencies.

jakubgs commented:
But I see this is more complex than just that. It needs more work.

OmarBasem commented:
@jakubgs Trying to run `make shell TARGET=gradle` or `make run-android` or `make android-clean` all fail with error: 
`inflating: /private/tmp/nix-build-aapt2-7.3.0-8691043.drv-0/tmp.i9ieU6xMHh/NOTICE  
patching sources
installing
error: builder for '/nix/store/wz30h0nd7vf31hjq36jqj68bdn1x8ifc-aapt2-7.3.0-8691043.drv' failed with exit code 12;
       last 8 log lines:
       > unpacking sources
       > Archive:  /nix/store/7nyznw5199fg5j1j6yawn1z17ks5kwmz-aapt2-7.3.0-8691043/aapt2-7.3.0-8691043-osx.jar
       >    creating: /private/tmp/nix-build-aapt2-7.3.0-8691043.drv-0/tmp.i9ieU6xMHh/META-INF/
       >   inflating: /private/tmp/nix-build-aapt2-7.3.0-8691043.drv-0/tmp.i9ieU6xMHh/META-INF/MANIFEST.MF
       >   inflating: /private/tmp/nix-build-aapt2-7.3.0-8691043.drv-0/tmp.i9ieU6xMHh/aapt2
       >   inflating: /private/tmp/nix-build-aapt2-7.3.0-8691043.drv-0/tmp.i9ieU6xMHh/NOTICE
       > patching sources
       > installing
       For full logs, run 'nix log /nix/store/wz30h0nd7vf31hjq36jqj68bdn1x8ifc-aapt2-7.3.0-8691043.drv'.
error: 1 dependencies of derivation '/nix/store/7fmq4lxn2sca1jz7wlc5gxa6knj6f3di-status-mobile-maven-deps.drv' failed to build
make: *** [shell] Error 1
`

jakubgs commented:
It means it needs more work. Also, please use triple backticks for large logs, otherwise the formatting is disgusting:
https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks#fenced-code-blocks

jakubgs committed:
add kotlin-android-extensions-runtime 1.6.20

Signed-off-by: Jakub Sokołowski <jakub@status.im>

jakubgs commented:
Some possibly related issues:

* https://github.com/RocketChat/Rocket.Chat.ReactNative/issues/4983
* https://github.com/software-mansion/react-native-reanimated/issues/280
* https://github.com/software-mansion/react-native-reanimated/issues/2045
* https://github.com/software-mansion/react-native-reanimated/issues/3334
* https://stackoverflow.com/questions/54858510/kotlin-compile-error-the-android-gradle-plugin-supports-only-kotlin-gradle-plu

jakubgs committed:
add kotlin-gradle-plugin 1.6.20

Signed-off-by: Jakub Sokołowski <jakub@status.im>

jakubgs committed:
add org.checkerframework:checker-qual:2.11.1

Signed-off-by: Jakub Sokołowski <jakub@status.im>

jakubgs committed:
upgrade aapt2 to 7.3.1-8691043

Signed-off-by: Jakub Sokołowski <jakub@status.im>

jakubgs commented:
I'm essentially switching between two errors. This one:
```
The Android Gradle plugin supports only Kotlin Gradle plugin version 1.5.20 and higher.
The following dependencies do not satisfy the required version:
root project 'StatusIm' -> org.jetbrains.kotlin:kotlin-gradle-plugin:1.3.50
```
Or this one:
```
The Android Gradle plugin supports only kotlin-android-extensions Gradle plugin version 1.6.20 and higher.
The following dependencies do not satisfy the required version:
root project 'StatusIm' -> org.jetbrains.kotlin:kotlin-gradle-plugin:1.5.20
```
@siddarthkay have you maybe dealt with this in your RN upgrade PR?

siddarthkay commented:
@OmarBasem : hey quick note on going to reanimated v3, there are breaking changes for usages of v1, I had briefly tried to capture them here https://github.com/status-im/status-mobile/issues/15418/
You must get rid of all v1 usage in code first and move that to v2 and only then you should upgrade this library :)

siddarthkay commented:
@OmarBasem : the change log on the official website is not helpful at all, what you must do is ensure we do not use any function that was part of v1 but absent from v2.
v3 is backwards compatible with v2 and after you upgrade you won't have to refactor those parts to get the app running.

jakubgs committed:
android: bump kotlin versions to 1.6.20

Signed-off-by: Jakub Sokołowski <jakub@status.im>

jakubgs commented:
Pushed some of my changes, but the fundamental issue is in the sub-projects that specify `1.4.31` as their Kotlin plugin version.

jakubgs commented:
@siddarthkay already created a tagged release of `react-native-hole-view` here at `v2.1.3-status`:
https://github.com/status-im/react-native-hole-view/tree/v2.1.3-status

Which does make it possible to override the Kotlin Gradle Plugin version, as seen here:
```gradle
    def kotlin_version = rootProject.ext.has('RNHV_kotlinVersion') ? rootProject.ext.get('RNHV_kotlinVersion') : project.properties['RNHV_kotlinVersion']
    def gradle_version = rootProject.ext.has('RNHV_gradleVersion') ? rootProject.ext.get('RNHV_gradleVersion') : project.properties['RNHV_gradleVersion']
```
https://github.com/status-im/react-native-hole-view/blob/v2.1.3-status/android/build.gradle#L3-L4
Which in theory should work.

jakubgs committed:
ugprade react-native-hole-view to v2.1.3-status

Signed-off-by: Jakub Sokołowski <jakub@status.im>

jakubgs commented:
Using `v2.1.3-status` release does work, but now we fail on a missing Gradle dependency:
```
Regenerating Nix files...
Found 40 sub-projects...
Found 699 direct dependencies...
Found 1045 dependency URLs...
 ! Failed to fetch: https://repo.maven.apache.org/maven2/org/ow2/asm/asm-bom/9.5/asm-bom-9.5.jar
```
Or specifically, a missing JAR/AAR for the POM:
https://repo.maven.apache.org/maven2/org/ow2/asm/asm-bom/9.5/

jakubgs committed:
nix: fix handling POMs without JARs

Signed-off-by: Jakub Sokołowski <jakub@status.im>

jakubgs committed:
nix: upgrade Android build tools to 30.0.3

```
* What went wrong:
Could not determine the dependencies of task ':app:lintVitalReportPr'.
> Failed to find Build Tools revision 30.0.3
```

Signed-off-by: Jakub Sokołowski <jakub@status.im>

jakubgs commented:
This is an issue that we already saw in the RN upgrade PR(https://github.com/status-im/status-mobile/pull/15203#issuecomment-1458869315):
```
* What went wrong:
Execution failed for task ':react-native-dialogs:extractReleaseAnnotations'.
> Could not resolve all files for configuration ':react-native-dialogs:detachedConfiguration1'.
   > Could not resolve com.android.tools.lint:lint-gradle:30.3.1.
     Required by:
         project :react-native-dialogs
      > No cached version of com.android.tools.lint:lint-gradle:30.3.1 available for offline mode.
      > No cached version of com.android.tools.lint:lint-gradle:30.3.1 available for offline mode.
```
For now I can add it manually, but it will disappear at next use of `make nix-update-gradle`.

jakubgs committed:
add com.android.tools.lint:lint-gradle:30.3.1 (TEMP)

Signed-off-by: Jakub Sokołowski <jakub@status.im>

jakubgs commented:
Wht the fuck is this:
```
Could not evaluate onlyIf predicate for task ':react-native-reanimated:downloadBoost'.
 > Unable to download file 'boost_1_63_0.tar.gz' in offline mode.
```

OmarBasem commented:
> You must get rid of all v1 usage in code first and move that to v2 and only then you should upgrade this library :)

@siddarthkay we can get rid of v1 code immediately no problem. I don't think we are using v1 code in `status-im2`

OmarBasem commented:
> Could not evaluate onlyIf predicate for task ':react-native-reanimated:downloadBoost'.
>  > Unable to download file 'boost_1_63_0.tar.gz' in offline mode.

@jakubgs related issue: https://github.com/software-mansion/react-native-reanimated/issues/3638

It seems that this `boost_1_63_0.tar.gz` file has been relocated

jakubgs commented:
Well, that issue says to upgrade to v2, which is probably not what we want:
```
547766 newfstatat(AT_FDCWD, "/home/jakubgs/work/status-mobile/node_modules/react-native-reanimated/android/build/downloads/boost_1_63_0.tar.gz", 0x7fcdf81e6f00, 0) = -1 ENOENT (No such file or directory)
```
This can be found in the strace log from Gradle. It's looking for it in:
```
node_modules/react-native-reanimated/android/build/downloads/boost_1_63_0.tar.gz
```
Wtf is this and why it's trying to download it in some custom way? No idea.

jakubgs commented:
Apparently this is only necessary when using RN older than 71:
```gradle
if (REACT_NATIVE_MINOR_VERSION < 71) {
    // You need to have following folders in this directory:
    //   - boost_1_63_0
    //   - double-conversion-1.1.6
    //   - folly-deprecate-dynamic-initializer
    //   - glog-0.3.5
    def dependenciesPath = System.getenv("REACT_NATIVE_DEPENDENCIES")

    // The Boost library is a very large download (>100MB).
    // If Boost is already present on your system, define the REACT_NATIVE_BOOST_PATH env variable
    // and the build will use that.
    def boostPath = dependenciesPath ?: System.getenv("REACT_NATIVE_BOOST_PATH")
```
https://github.com/software-mansion/react-native-reanimated/blob/3.1.0/android/build.gradle#L705-L716

Now that's some dumb shit. Do they just expect people to download those manually?

jakubgs commented:
They actually download them:
```gradle
    if (REACT_NATIVE_MINOR_VERSION < 71) {
        extractAARHeaders.dependsOn(prepareThirdPartyNdkHeaders)
        extractSOFiles.dependsOn(prepareThirdPartyNdkHeaders)

        nativeBuildDependsOn(prepareThirdPartyNdkHeaders)
```
```gradle
    task prepareThirdPartyNdkHeaders(dependsOn:[
        downloadNdkBuildDependencies,
```
```gradle
    task downloadBoost(dependsOn: resolveBoost, type: Download) {
        def transformedVersion = BOOST_VERSION.replace("_", ".")
        def artifactLocalName = "boost_${BOOST_VERSION}.tar.gz"
        def srcUrl = "https://boostorg.jfrog.io/artifactory/main/release/${transformedVersion}/source/boost_${BOOST_VERSION}.tar.gz"
        if (REACT_NATIVE_MINOR_VERSION < 69) {
            srcUrl = "https://github.com/react-native-community/boost-for-react-native/releases/download/v${transformedVersion}-0/boost_${BOOST_VERSION}.tar.gz"
        }
```
But since the Nix build is pure there is no network for it to do that.

Why do they re-invent the wheel and do their own dependency management? Who thought this was a good idea?
Now we might have to write some custom Nix derivations to download this and then patch the module to make this work.

Or we could actually do the React Native upgrade...

siddarthkay commented:
@jakubgs : I am resuming work on the upgrade this week, should be fun, will keep you posted.

OmarBasem commented:
>  if (REACT_NATIVE_MINOR_VERSION < 71) { 

Oh I see..

OmarBasem commented:
Fine we can do the RN upgrade then

OmarBasem commented:
Dependency: https://github.com/status-im/status-mobile/pull/15203
